README.md
# RAWRTC

[![CircleCI build status][circleci-badge]][circleci-url]
[![Travis CI build status][travis-ci-badge]][travis-ci-url]
[![Join our chat on Gitter][gitter-icon]][gitter]

A WebRTC and ORTC library with a small footprint.

## Features

The following list represents all features that are planned for RAWRTC.
Features with a check mark are already implemented.

* ICE [[draft-ietf-ice-rfc-5245bis-08]][ice]
  - [x] Trickle ICE [[draft-ietf-ice-trickle-07]][trickle-ice]
  - [x] IPv4
  - [x] IPv6
  - [x] UDP
  - [ ] TCP
* STUN [[RFC 5389]][stun]
  - [x] UDP
  - [ ] TCP
  - [ ] TLS over TCP
  - [ ] DTLS over UDP [[RFC 7350]][stun-turn-dtls]
* TURN [[RFC 5766]][turn]
  - [ ] UDP
  - [ ] TCP
  - [ ] TLS over TCP
  - [ ] DTLS over UDP [[RFC 7350]][stun-turn-dtls]
* Data Channel
  - [x] DCEP [[draft-ietf-rtcweb-data-protocol-09]][dcep]
  - [x] SCTP-based [[draft-ietf-rtcweb-data-channel-13]][sctp-dc]
* API
  - [x] WebRTC C-API based on the [W3C WebRTC API][w3c-webrtc] and
    [[draft-ietf-rtcweb-jsep-24]][jsep]
  - [x] ORTC C-API based on the [W3C CG ORTC API][w3c-ortc]

## FAQ

1. *Does this use multiple threads?*

   No, this library is single-threaded and uses an event loop underneath.

2. *Can I use it in a threaded environment?*

   Yes. Just make sure you're always calling it from the same thread the event
   loop is running on, or either [lock/unlock the event loop thread][re-lock]
   or [use the message queues provided by re][re-mqueue] to exchange data with
   the event loop thread. However, it is important that you only run one *re*
   event loop in one thread.

3. *I only want a data channel implementation for my SFU!*

   Check out [RAWRTCDC][rawrtcdc].

## Prerequisites

The following tools are required:

* [git][git]
* [ninja][ninja] >= 1.5
* [meson][meson] >= 0.48.0
* pkg-config (`pkgconf` for newer FreeBSD versions)
* SSL development libraries (`libssl-dev` on Debian, `openssl` on OSX and
  FreeBSD)

## Build

```bash
cd <path-to-rawrtc>
mkdir build
meson build
cd build
ninja
```

## Run

RAWRTC provides a lot of tools that can be used for quick testing purposes and
to get started. Let's go through them one by one. If you just want to check out
data channels and browser interoperation, skip to the
[`peer-connection` tool section](#peer-connection) which uses the WebRTC API or
to the [`data-channel-sctp` tool section](#data-channel-sctp) which uses the
ORTC API.

Most of the tools have required or optional arguments which are shared among
tools. Below is a description for the various shared arguments:

#### offering

Whether the peer is going to create an offer. Provide `1` to create an offer
immediately or `0` to create an answer once the remote offer has been
processed.

Only used by WebRTC API tools.

#### ice-role

Determines the ICE role to be used by the ICE transport, where `0` means
*controlled* and `1` means *controlling*.

Only used by ORTC API tools.

#### sctp-port

The port number the internal SCTP stack is supposed to use. Defaults to `5000`.

Note: It doesn't matter which port you choose unless you want to be able to
debug SCTP messages. In this case, it's easier to distinguish the peers by
their port numbers.

Only used by ORTC API tools.

#### ice-candidate-type

If supplied, one or more specific ICE candidate types will be enabled and all
other ICE candidate types will be disabled. Can be one of the following
strings:

* *host*
* *srflx*
* *prflx*
* *relay*

Note that this has no effect on the gathering policy. The candidates will be
gathered but they will simply be ignored by the tool.

If not supplied, all ICE candidate types are enabled.

### ice-gatherer

API: ORTC

The ICE gatherer tool gathers and prints ICE candidates. Once gathering is
complete, the tool exits.

Usage:

    ice-gatherer

### ice-transport-loopback

API: ORTC

The ICE transport loopback tool starts two ICE transport instances which
establish an ICE connection. Once you see the following line for both clients
*A* and *B*, the ICE connection has been established:

    (<client>) ICE transport state: connected

Usage:

    ice-transport-loopback [<ice-candidate-type> ...]

### dtls-transport-loopback

API: ORTC

The DTLS transport loopback tool starts two DTLS transport instances which
work on top of an established ICE transport connection.

To verify that the DTLS connection establishes, wait for the following line for
both clients *A* and *B*:

    (<client>) DTLS transport state change: connected

Usage:

    dtls-transport-loopback [<ice-candidate-type> ...]

### sctp-redirect-transport

API: ORTC

The SCTP redirect transport tool starts an SCTP redirect transport on top of an
established DTLS transport to relay SCTP messages from and to a third party.
This tool has been developed to be able to test data channel implementations
without having to write the required DTLS and ICE stacks. An example of such a
testing tool is [dctt][dctt] which uses the kernel SCTP stack of FreeBSD.

Building:

This tool is not built by default. You can enable building it in the following
way:

```bash
cd <path-to-rawrtc>/build
meson configure -Dsctp_redirect_transport=true
ninja
```

Usage:

    sctp-redirect-transport <0|1 (ice-role)> <redirect-ip> <redirect-port>
                            [<sctp-port>] [<maximum-message-size>]
                            [<ice-candidate-type> ...]

Special arguments:

* `redirect-ip`: The IP address on which the external SCTP stack is listening.
* `redirect-port` The port on which the external SCTP stack is listening.
* `maximum-message-size`: The maximum message size of a data channel message
  the external SCTP stack is able to handle. `0` indicates that messages of
  arbitrary size can be handled. Defaults to `0`.

### data-channel-sctp-loopback

API: ORTC

The data channel SCTP loopback tool creates several data channels on top of an
abstracted SCTP data transport. As soon as a data channel is open, a message
will be sent to the other peer. Furthermore, another message will be sent on a
specific channel after a brief timeout.

To verify that a data channels opens, wait for the following line:

    (<client>) Data channel open: <channel-label>
    
The tool will send some large (16 MiB) test data to the other peer depending on
the ICE role. We are able to do this because RAWRTC handles data channel
messages correctly and does not have a maximum message size limitation compared
to most other implementations (check out
[this article][demystifying-webrtc-dc-size-limit] for a detailed explanation).

Usage:

    data-channel-sctp-loopback [<ice-candidate-type> ...]

### data-channel-sctp

API: ORTC

The data channel SCTP tool creates several data channels on top of an
abstracted SCTP data transport:

* A pre-negotiated data channel with the label `cat-noises` and the id `0`
  that is reliable and ordered. In the WebRTC JS API, the channel would be
  created by invoking:
   
   ```js
   peerConnection.createDataChannel('cat-noises', {
       ordered: true,
       id: 0
   });
   ```

* A data channel with the label `bear-noises` that is reliable but unordered.
  In the WebRTC JS API, the channel would be created by invoking:
   
   ```js
   peerConnection.createDataChannel('bear-noises', {
       ordered: false,
       maxRetransmits: 0
   });
   ```

To establish a connection with another peer, the following procecure must be
followed:

1. The JSON blob after `Local Parameters:` must be pasted into the other peer
   you want to establish a connection with. This can be either a browser
   instance that uses the 
   [WebRTC-ORTC browser example tool][webrtc-ortc-example] or another instance
   of this tool.

2. The other peer's local parameters in form of a JSON blob must be pasted into
   this tool's instance.

3. Once you've pasted the local parameters into each other's instance, the peer
   connection can be established by pressing *Enter* in both instances (press
   the *Start* button in the browser).

The tool will send some test data to the other peer depending on the ICE role.
However, the browser tool behaves a bit differently. Check the log output of
the tool instances (console output in the browser) to see what data has been
sent and whether it has been received successfully.

In the browser, you can use the created data channels by accessing
`peer.dc['<channel-name>']`, for example:

```js
peer.dc['example-channel'].send('RAWR!')
```

Usage:

    data-channel-sctp <0|1 (ice-role)> [<sctp-port>] [<ice-candidate-type> ...]

### data-channel-sctp-streamed

API: ORTC

The data channel SCTP streamed tool is the counterpart to the *normal* data
channel SCTP tool but uses the streaming mode. **Be aware this tool and the
streaming mode is currently experimental and incomplete.**

The necessary peer connection establishment steps are identical to the ones
described for the [data-channel-sctp](#data-channel-sctp) tool.

Usage:

    data-channel-sctp-streamed <0|1 (ice-role)> [<sctp-port>]
                               [<ice-candidate-type> ...]

### data-channel-sctp-echo

API: ORTC

The data channel SCTP echo tool behaves just like any other echo server: It
echoes received data on any data channel back to the sender.

The necessary peer connection establishment steps are identical to the ones
described for the [data-channel-sctp](#data-channel-sctp) tool.

Usage:

    data-channel-sctp-echo <0|1 (ice-role)> [<sctp-port>]
                           [<ice-candidate-type> ...]
    
### data-channel-sctp-throughput

API: ORTC

The data channel SCTP throughput tool allows you to test throughput by sending
one or more message. It will report the amount of seconds elapsed and the
throughput in Mbit/s.

The necessary peer connection establishment steps are identical to the ones
described for the [data-channel-sctp](#data-channel-sctp) tool. However,
be aware that this tool has no browser counterpart at the moment, so it only
makes sense to use two instances of this tool for throughput testing.

Usage:

    data-channel-sctp-throughput <0|1 (ice-role)> <message-size> [<n-times>]
                                 [<sctp-port>] [<ice-candidate-type> ...]

Special arguments:

* `message-size`: Is the message size in bytes used for throughput testing. The
  controlling peer will determine the message size for both peers, so this
  argument is being ignored for the controlled peer.
* `n-times`: Is the amount of times the message will be sent. Again, this
  value is being ignored for the controlled peer.

### peer-connection

API: WebRTC

The peer connection tool creates a peer connection instance and several data
channels:

* A pre-negotiated data channel with the label `cat-noises` and the id `0`
  that is reliable and ordered. In the JS API, the channel would be created
  by invoking:
   
   ```js
   peerConnection.createDataChannel('cat-noises', {
       ordered: true,
       id: 0
   });
   ```

* A data channel with the label `bear-noises` that is reliable but unordered.
  In the WebRTC JS API, the channel would be created by invoking:
   
   ```js
   peerConnection.createDataChannel('bear-noises', {
       ordered: false,
       maxRetransmits: 0
   });
   ```

To establish a connection with another peer, the following procecure must be
followed:

1. If the peer is taking the *offering* role, the generated JSON blob that
   contains the *offer SDP* must be pasted into the other peer you want to
   establish a connection with. This can be either a browser instance that uses
   the [WebRTC browser example tool][webrtc-example] or another instance of
   this tool. In case it is a browser instance, press the *Start* button and
   paste the data directly into the text area below
   `Paste remote description:`. In case it is another instance of this tool,
   paste the data into the other peer's console and press *Enter*.

2. The peer who takes the *answering* role now generates a JSON blob as well
   that contains the *answer SDP*. It must be pasted into the other browser
   instance or tool instance as described in the previous step.

3. The peer connection should be established automatically once *offer* and
   *answer* have been exchanged and applied.

The tool will send some test data to the other peer depending on whether or not
it took the *offering* role. However, the browser tool behaves a bit
differently. Check the log output of the tool instances (in the browser, either
open the console log or check out the live log on the right side) to see what
data has been sent and whether it has been received successfully.

In the browser, you can use the created data channels by accessing
`pc.dcs['<channel-name>']` in the console log, for example:

```js
pc.dcs['cat-noises'].send('RAWR!')
```

Usage:

    peer-connection <0|1 (offering)> [<ice-candidate-type> ...]

## Contributing

When creating a pull request, it is recommended to run `format-all.sh` to
apply a consistent code style.



[circleci-badge]: https://circleci.com/gh/rawrtc/rawrtc.svg?style=shield
[circleci-url]: https://circleci.com/gh/rawrtc/rawrtc
[travis-ci-badge]: https://travis-ci.org/rawrtc/rawrtc.svg?branch=master
[travis-ci-url]: https://travis-ci.org/rawrtc/rawrtc
[gitter]: https://gitter.im/rawrtc/Lobby
[gitter-icon]: https://badges.gitter.im/rawrtc/Lobby.svg

[ice]: https://tools.ietf.org/html/draft-ietf-ice-rfc5245bis-08

CHANGELOG.md
# Changelog

## [0.5.2] (2021-09-20)

* Fix usrsctp dependency (#154)

## [0.5.1] (2019-08-15)

* Re-enable peer reflexive candidates (#141)

## [0.5.0] (2019-08-15)

* Fix BoringSSL compatibility (#139)
* Calculate ICE candidate priority (#140)
* Use [upstream usrsctp](https://github.com/sctplab/usrsctp/)
* Expose more transport parameters (#146)
  - Add getter/setter for send/receiver buffer length
  - Add getter/setter for congestion control algorithm
  - Add getter/setter for MTU

## [0.4.0] (2019-03-19)

* Replace CMake with Meson build system
  - This finally allowed us to get rid of the dependencies script :tada:
* Internal restructuring of source files

## [0.3.0] (2019-03-02)

* Split the stack into three separate entities (major)
  - RAWRTC (this repository) contains the WebRTC/ORTC API, the ICE/DTLS stack
    and bindings to the data channel implementation.
  - [RAWRTCDC](https://github.com/rawrtc/rawrtc-data-channel) contains the data
    channel implementation
  - [RAWRTCC](https://github.com/rawrtc/rawrtc-common) contains common
    functionality required by both RAWRTC and RAWRTCDC
* Added a gathering timeout
* Fixed various issues with the ICE transport states

## [0.2.2] (2018-04-14)

* Fixed parsing the DTLS role in the peer connection API

## [0.2.1] (2018-02-26)

* Fixed missing cast in examples leading to issues on specific platforms (such
  as ARM)

## [0.2.0] (2018-02-12)

* Initial release of RAWRTC



[0.5.2]: https://github.com/rawrtc/rawrtc/compare/v0.5.1...v0.5.2
[0.5.1]: https://github.com/rawrtc/rawrtc/compare/v0.5.0...v0.5.1
[0.5.0]: https://github.com/rawrtc/rawrtc/compare/v0.4.0...v0.5.0
[0.4.0]: https://github.com/rawrtc/rawrtc/compare/v0.3.0...v0.4.0
[0.3.0]: https://github.com/rawrtc/rawrtc/compare/v0.2.2...v0.3.0
[0.2.2]: https://github.com/rawrtc/rawrtc/compare/v0.2.1...v0.2.2
[0.2.1]: https://github.com/rawrtc/rawrtc/compare/v0.2.0...v0.2.1
[0.2.0]: https://github.com/rawrtc/rawrtc/compare/bd9d1ef15d008fdc24b4d5e3158e775a03ffec16...v0.2.0


htdocs/ortc/sdp.js
/*
 * Copyright Philipp Hancke
 * License: MIT
 * Source: https://github.com/fippo/sdp
 *
 * Extended by Lennart Grahl
 */

/* eslint-env node */
'use strict';

// SDP helpers.
var SDPUtils = {};

// Generate an alphanumeric identifier for cname or mids.
// TODO: use UUIDs instead? https://gist.github.com/jed/982883
SDPUtils.generateIdentifier = function() {
  return Math.random().toString(36).substr(2, 10);
};

// The RTCP CNAME used by all peerconnections from the same JS.
SDPUtils.localCName = SDPUtils.generateIdentifier();

// Splits SDP into lines, dealing with both CRLF and LF.
SDPUtils.splitLines = function(blob) {
  return blob.trim().split('\n').map(function(line) {
    return line.trim();
  });
};
// Splits SDP into sessionpart and mediasections. Ensures CRLF.
SDPUtils.splitSections = function(blob) {
  var parts = blob.split('\nm=');
  return parts.map(function(part, index) {
    return (index > 0 ? 'm=' + part : part).trim() + '\r\n';
  });
};

// Returns lines that start with a certain prefix.
SDPUtils.matchPrefix = function(blob, prefix) {
  return SDPUtils.splitLines(blob).filter(function(line) {
    return line.indexOf(prefix) === 0;
  });
};

// Parses an ICE candidate line. Sample input:
// candidate:702786350 2 udp 41819902 8.8.8.8 60769 typ relay raddr 8.8.8.8
// rport 55996"
SDPUtils.parseCandidate = function(line) {
  var parts;
  // Parse both variants.
  if (line.indexOf('a=candidate:') === 0) {
    parts = line.substring(12).split(' ');
  } else {
    parts = line.substring(10).split(' ');
  }

  var candidate = {
    foundation: parts[0],
    component: parts[1],
    protocol: parts[2].toLowerCase(),
    priority: parseInt(parts[3], 10),
    ip: parts[4],
    port: parseInt(parts[5], 10),
    // skip parts[6] == 'typ'
    type: parts[7]
  };

  for (var i = 8; i < parts.length; i += 2) {
    switch (parts[i]) {
      case 'raddr':
        candidate.relatedAddress = parts[i + 1];
        break;
      case 'rport':
        candidate.relatedPort = parseInt(parts[i + 1], 10);
        break;
      case 'tcptype':
        candidate.tcpType = parts[i + 1];
        break;
      default: // Unknown extensions are silently ignored.
        break;
    }
  }
  return candidate;
};

// Translates a candidate object into SDP candidate attribute.
SDPUtils.writeCandidate = function(candidate) {
  var sdp = [];
  sdp.push(candidate.foundation);
  sdp.push(candidate.component);
  sdp.push(candidate.protocol.toUpperCase());
  sdp.push(candidate.priority);
  sdp.push(candidate.ip);
  sdp.push(candidate.port);

  var type = candidate.type;
  sdp.push('typ');
  sdp.push(type);
  if (type !== 'host' && candidate.relatedAddress &&
      candidate.relatedPort) {
    sdp.push('raddr');
    sdp.push(candidate.relatedAddress); // was: relAddr
    sdp.push('rport');
    sdp.push(candidate.relatedPort); // was: relPort
  }
  if (candidate.tcpType && candidate.protocol.toLowerCase() === 'tcp') {
    sdp.push('tcptype');
    sdp.push(candidate.tcpType);
  }
  return 'candidate:' + sdp.join(' ');
};

// Parses an rtpmap line, returns RTCRtpCoddecParameters. Sample input:
// a=rtpmap:111 opus/48000/2
SDPUtils.parseRtpMap = function(line) {
  var parts = line.substr(9).split(' ');
  var parsed = {
    payloadType: parseInt(parts.shift(), 10) // was: id
  };

  parts = parts[0].split('/');

  parsed.name = parts[0];
  parsed.clockRate = parseInt(parts[1], 10); // was: clockrate
  // was: channels
  parsed.numChannels = parts.length === 3 ? parseInt(parts[2], 10) : 1;
  return parsed;
};

// Generate an a=rtpmap line from RTCRtpCodecCapability or
// RTCRtpCodecParameters.
SDPUtils.writeRtpMap = function(codec) {
  var pt = codec.payloadType;
  if (codec.preferredPayloadType !== undefined) {
    pt = codec.preferredPayloadType;
  }
  return 'a=rtpmap:' + pt + ' ' + codec.name + '/' + codec.clockRate +
      (codec.numChannels !== 1 ? '/' + codec.numChannels : '') + '\r\n';
};

// Parses an a=extmap line (headerextension from RFC 5285). Sample input:
// a=extmap:2 urn:ietf:params:rtp-hdrext:toffset
SDPUtils.parseExtmap = function(line) {
  var parts = line.substr(9).split(' ');
  return {
    id: parseInt(parts[0], 10),
    uri: parts[1]
  };
};

// Generates a=extmap line from RTCRtpHeaderExtensionParameters or
// RTCRtpHeaderExtension.
SDPUtils.writeExtmap = function(headerExtension) {
  return 'a=extmap:' + (headerExtension.id || headerExtension.preferredId) +
      ' ' + headerExtension.uri + '\r\n';
};

// Parses an ftmp line, returns dictionary. Sample input:
// a=fmtp:96 vbr=on;cng=on
// Also deals with vbr=on; cng=on
SDPUtils.parseFmtp = function(line) {
  var parsed = {};
  var kv;
  var parts = line.substr(line.indexOf(' ') + 1).split(';');
  for (var j = 0; j < parts.length; j++) {
    kv = parts[j].trim().split('=');
    parsed[kv[0].trim()] = kv[1];
  }
  return parsed;
};

// Generates an a=ftmp line from RTCRtpCodecCapability or RTCRtpCodecParameters.
SDPUtils.writeFmtp = function(codec) {
  var line = '';
  var pt = codec.payloadType;
  if (codec.preferredPayloadType !== undefined) {
    pt = codec.preferredPayloadType;
  }
  if (codec.parameters && Object.keys(codec.parameters).length) {
    var params = [];
    Object.keys(codec.parameters).forEach(function(param) {
      params.push(param + '=' + codec.parameters[param]);
    });
    line += 'a=fmtp:' + pt + ' ' + params.join(';') + '\r\n';
  }
  return line;
};

// Parses an rtcp-fb line, returns RTCPRtcpFeedback object. Sample input:
// a=rtcp-fb:98 nack rpsi
SDPUtils.parseRtcpFb = function(line) {
  var parts = line.substr(line.indexOf(' ') + 1).split(' ');
  return {
    type: parts.shift(),
    parameter: parts.join(' ')
  };
};
// Generate a=rtcp-fb lines from RTCRtpCodecCapability or RTCRtpCodecParameters.
SDPUtils.writeRtcpFb = function(codec) {
  var lines = '';
  var pt = codec.payloadType;
  if (codec.preferredPayloadType !== undefined) {
    pt = codec.preferredPayloadType;
  }
  if (codec.rtcpFeedback && codec.rtcpFeedback.length) {
    // FIXME: special handling for trr-int?
    codec.rtcpFeedback.forEach(function(fb) {
      lines += 'a=rtcp-fb:' + pt + ' ' + fb.type +
          (fb.parameter && fb.parameter.length ? ' ' + fb.parameter : '') +
          '\r\n';
    });
  }
  return lines;
};

// Parses an RFC 5576 ssrc media attribute. Sample input:
// a=ssrc:3735928559 cname:something
SDPUtils.parseSsrcMedia = function(line) {
  var sp = line.indexOf(' ');
  var parts = {
    ssrc: parseInt(line.substr(7, sp - 7), 10)
  };
  var colon = line.indexOf(':', sp);
  if (colon > -1) {
    parts.attribute = line.substr(sp + 1, colon - sp - 1);
    parts.value = line.substr(colon + 1);
  } else {
    parts.attribute = line.substr(sp + 1);
  }
  return parts;
};

// Extracts SCTP capabilities from SDP media section or sessionpart.
SDPUtils.getSctpCapabilities = function(mediaSection, sessionpart) {
  var lines = SDPUtils.splitLines(mediaSection);
  // Search in session part, too.
  lines = lines.concat(SDPUtils.splitLines(sessionpart));
  var maxMessageSize = lines.filter(function(line) {
    return line.indexOf('a=max-message-size:') === 0;
  });
  // TODO: Use 65536 once Firefox has disabled PPID-based fragmentation
  //       see: https://lgrahl.de/articles/demystifying-webrtc-dc-size-limit.html
  maxMessageSize = maxMessageSize.length ? parseInt(maxMessageSize[0].substr(19)) : 16384;
  return {
    maxMessageSize: maxMessageSize
  };
};

// Serializes SCTP capabilities to SDP.
SDPUtils.writeSctpCapabilities = function(capabilities) {
  return 'a=max-message-size:' + capabilities.maxMessageSize + '\r\n'; // (03)
};

// Extracts SCTP port from SDP media section or sessionpart.
SDPUtils.getSctpPort = function(mediaSection, sessionpart) {
  var lines = SDPUtils.splitLines(mediaSection);
  // Search in session part, too.
  lines = lines.concat(SDPUtils.splitLines(sessionpart));
  var port = lines.filter(function(line) {
    return line.indexOf('a=sctp-port:') === 0;
  });
  port = port.length ? parseInt(port[0].substr(12), 10) : 5000;
  return port;
};

// Serializes SCTP port to SDP.
SDPUtils.writeSctpPort = function(port) {
  // TODO: Enable (chromium can't cope with it)
  // return 'a=sctp-port:' + (port ? port : 5000) + '\r\n'; // (03)
  return '';
};

// Extracts DTLS parameters from SDP media section or sessionpart.
// FIXME: for consistency with other functions this should only
//   get the fingerprint line as input. See also getIceParameters.
SDPUtils.getDtlsParameters = function(mediaSection, sessionpart) {
  var lines = SDPUtils.splitLines(mediaSection);
  // Search in session part, too.
  lines = lines.concat(SDPUtils.splitLines(sessionpart));
  var fpLine = lines.filter(function(line) {
    return line.indexOf('a=fingerprint:') === 0;
  })[0].substr(14);
  // Note: a=setup line is ignored since we use the 'auto' role.
  var dtlsParameters = {
    role: 'auto',
    fingerprints: [{
      algorithm: fpLine.split(' ')[0],
      value: fpLine.split(' ')[1]
    }]
  };
  return dtlsParameters;
};

// Serializes DTLS parameters to SDP.
SDPUtils.writeDtlsParameters = function(params, setupType) {
  var sdp = 'a=setup:' + setupType + '\r\n';
  params.fingerprints.forEach(function(fp) {
    sdp += 'a=fingerprint:' + fp.algorithm + ' ' + fp.value + '\r\n';
  });
  return sdp;
};
// Parses ICE information from SDP media section or sessionpart.
// FIXME: for consistency with other functions this should only
//   get the ice-ufrag and ice-pwd lines as input.
SDPUtils.getIceParameters = function(mediaSection, sessionpart) {
  var lines = SDPUtils.splitLines(mediaSection);
  // Search in session part, too.
  lines = lines.concat(SDPUtils.splitLines(sessionpart));
  var iceParameters = {
    usernameFragment: lines.filter(function(line) {
      return line.indexOf('a=ice-ufrag:') === 0;
    })[0].substr(12),
    password: lines.filter(function(line) {
      return line.indexOf('a=ice-pwd:') === 0;
    })[0].substr(10)
  };
  return iceParameters;
};

// Serializes ICE parameters to SDP.
SDPUtils.writeIceParameters = function(params) {
  return 'a=ice-ufrag:' + params.usernameFragment + '\r\n' +
      'a=ice-pwd:' + params.password + '\r\n';
};

// Parses the SDP media section and returns RTCRtpParameters.
SDPUtils.parseRtpParameters = function(mediaSection) {
  var description = {
    codecs: [],
    headerExtensions: [],
    fecMechanisms: [],
    rtcp: []
  };
  var lines = SDPUtils.splitLines(mediaSection);
  var mline = lines[0].split(' ');
  for (var i = 3; i < mline.length; i++) { // find all codecs from mline[3..]
    var pt = mline[i];
    var rtpmapline = SDPUtils.matchPrefix(
        mediaSection, 'a=rtpmap:' + pt + ' ')[0];
    if (rtpmapline) {
      var codec = SDPUtils.parseRtpMap(rtpmapline);
      var fmtps = SDPUtils.matchPrefix(
          mediaSection, 'a=fmtp:' + pt + ' ');
      // Only the first a=fmtp:<pt> is considered.
      codec.parameters = fmtps.length ? SDPUtils.parseFmtp(fmtps[0]) : {};
      codec.rtcpFeedback = SDPUtils.matchPrefix(
          mediaSection, 'a=rtcp-fb:' + pt + ' ')
          .map(SDPUtils.parseRtcpFb);
      description.codecs.push(codec);
      // parse FEC mechanisms from rtpmap lines.
      switch (codec.name.toUpperCase()) {
        case 'RED':
        case 'ULPFEC':
          description.fecMechanisms.push(codec.name.toUpperCase());
          break;
        default: // only RED and ULPFEC are recognized as FEC mechanisms.
          break;
      }
    }
  }
  SDPUtils.matchPrefix(mediaSection, 'a=extmap:').forEach(function(line) {
    description.headerExtensions.push(SDPUtils.parseExtmap(line));
  });
  // FIXME: parse rtcp.
  return description;
};

// Generates parts of the SDP media section describing the capabilities /
// parameters.
SDPUtils.writeRtpDescription = function(kind, caps) {
  var sdp = '';

  // Build the mline.
  sdp += 'm=' + kind + ' ';
  sdp += caps.codecs.length > 0 ? '9' : '0'; // reject if no codecs.
  sdp += ' UDP/TLS/RTP/SAVPF ';
  sdp += caps.codecs.map(function(codec) {
        if (codec.preferredPayloadType !== undefined) {
          return codec.preferredPayloadType;
        }
        return codec.payloadType;
      }).join(' ') + '\r\n';

  sdp += 'c=IN IP4 0.0.0.0\r\n';
  sdp += 'a=rtcp:9 IN IP4 0.0.0.0\r\n';

  // Add a=rtpmap lines for each codec. Also fmtp and rtcp-fb.
  caps.codecs.forEach(function(codec) {
    sdp += SDPUtils.writeRtpMap(codec);
    sdp += SDPUtils.writeFmtp(codec);
    sdp += SDPUtils.writeRtcpFb(codec);
  });
  var maxptime = 0;
  caps.codecs.forEach(function(codec) {
    if (codec.maxptime > maxptime) {
      maxptime = codec.maxptime;
    }
  });
  if (maxptime > 0) {
    sdp += 'a=maxptime:' + maxptime + '\r\n';
  }
  sdp += 'a=rtcp-mux\r\n';

  caps.headerExtensions.forEach(function(extension) {
    sdp += SDPUtils.writeExtmap(extension);
  });
  // FIXME: write fecMechanisms.
  return sdp;
};

// Parses the SDP media section and returns an array of
// RTCRtpEncodingParameters.
SDPUtils.parseRtpEncodingParameters = function(mediaSection) {
  var encodingParameters = [];
  var description = SDPUtils.parseRtpParameters(mediaSection);
  var hasRed = description.fecMechanisms.indexOf('RED') !== -1;
  var hasUlpfec = description.fecMechanisms.indexOf('ULPFEC') !== -1;

  // filter a=ssrc:... cname:, ignore PlanB-msid
  var ssrcs = SDPUtils.matchPrefix(mediaSection, 'a=ssrc:')
      .map(function(line) {
        return SDPUtils.parseSsrcMedia(line);
      })
      .filter(function(parts) {
        return parts.attribute === 'cname';
      });
  var primarySsrc = ssrcs.length > 0 && ssrcs[0].ssrc;
  var secondarySsrc;

  var flows = SDPUtils.matchPrefix(mediaSection, 'a=ssrc-group:FID')
      .map(function(line) {
        var parts = line.split(' ');
        parts.shift();
        return parts.map(function(part) {
          return parseInt(part, 10);
        });
      });
  if (flows.length > 0 && flows[0].length > 1 && flows[0][0] === primarySsrc) {
    secondarySsrc = flows[0][1];
  }

  description.codecs.forEach(function(codec) {
    if (codec.name.toUpperCase() === 'RTX' && codec.parameters.apt) {
      var encParam = {
        ssrc: primarySsrc,
        codecPayloadType: parseInt(codec.parameters.apt, 10),
        rtx: {
          payloadType: codec.payloadType,
          ssrc: secondarySsrc
        }
      };
      encodingParameters.push(encParam);
      if (hasRed) {
        encParam = JSON.parse(JSON.stringify(encParam));
        encParam.fec = {
          ssrc: secondarySsrc,
          mechanism: hasUlpfec ? 'red+ulpfec' : 'red'
        };
        encodingParameters.push(encParam);
      }
    }
  });
  if (encodingParameters.length === 0 && primarySsrc) {
    encodingParameters.push({
      ssrc: primarySsrc
    });
  }

  // we support both b=AS and b=TIAS but interpret AS as TIAS.
  var bandwidth = SDPUtils.matchPrefix(mediaSection, 'b=');
  if (bandwidth.length) {
    if (bandwidth[0].indexOf('b=TIAS:') === 0) {
      bandwidth = parseInt(bandwidth[0].substr(7), 10);
    } else if (bandwidth[0].indexOf('b=AS:') === 0) {
      bandwidth = parseInt(bandwidth[0].substr(5), 10);
    }
    encodingParameters.forEach(function(params) {
      params.maxBitrate = bandwidth;
    });
  }
  return encodingParameters;
};

SDPUtils.writeSessionBoilerplate = function() {
  // FIXME: sess-id should be an NTP timestamp.
  return 'v=0\r\n' +
      'o=thisisadapterortc 8169639915646943137 2 IN IP4 127.0.0.1\r\n' +
      's=-\r\n' +
      't=0 0\r\n';
};

SDPUtils.writeMediaSection = function(transceiver, caps, type, stream) {
  var sdp = SDPUtils.writeRtpDescription(transceiver.kind, caps);

  // Map ICE parameters (ufrag, pwd) to SDP.
  sdp += SDPUtils.writeIceParameters(
      transceiver.iceGatherer.getLocalParameters());

  // Map DTLS parameters to SDP.
  sdp += SDPUtils.writeDtlsParameters(
      transceiver.dtlsTransport.getLocalParameters(),
      type === 'offer' ? 'actpass' : 'active');

  sdp += 'a=mid:' + transceiver.mid + '\r\n';

  if (transceiver.rtpSender && transceiver.rtpReceiver) {
    sdp += 'a=sendrecv\r\n';
  } else if (transceiver.rtpSender) {
    sdp += 'a=sendonly\r\n';
  } else if (transceiver.rtpReceiver) {
    sdp += 'a=recvonly\r\n';
  } else {
    sdp += 'a=inactive\r\n';
  }

  // FIXME: for RTX there might be multiple SSRCs. Not implemented in Edge yet.
  if (transceiver.rtpSender) {
    var msid = 'msid:' + stream.id + ' ' +
        transceiver.rtpSender.track.id + '\r\n';
    sdp += 'a=' + msid;
    sdp += 'a=ssrc:' + transceiver.sendEncodingParameters[0].ssrc +
        ' ' + msid;
  }
  // FIXME: this should be written by writeRtpDescription.
  sdp += 'a=ssrc:' + transceiver.sendEncodingParameters[0].ssrc +
      ' cname:' + SDPUtils.localCName + '\r\n';
  return sdp;
};

// Gets the direction from the mediaSection or the sessionpart.
SDPUtils.getDirection = function(mediaSection, sessionpart) {
  // Look for sendrecv, sendonly, recvonly, inactive, default to sendrecv.
  var lines = SDPUtils.splitLines(mediaSection);
  for (var i = 0; i < lines.length; i++) {
    switch (lines[i]) {
      case 'a=sendrecv':
      case 'a=sendonly':
      case 'a=recvonly':
      case 'a=inactive':
        return lines[i].substr(2);
      default:
        // FIXME: What should happen here?
    }
  }
  if (sessionpart) {
    return SDPUtils.getDirection(sessionpart);
  }
  return 'sendrecv';
};
